File: backend/app/routers/goals.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from datetime import datetime, timezone

from app.coaching import overlays as coach_overlays
from app.core.db import get_db
from app.core.security import get_current_user
from app.models.calorie_goal import CalorieGoal
from app.models.coaching_baseline import CoachingBaseline
from app.models.user import User
from app.schemas.goal import CalorieGoalAutoSuggestion, CalorieGoalManualCreate, CalorieGoalRead
from app.services import goal_service

router = APIRouter(prefix="/goals/calorie", tags=["goals"])


def _require_profile(current_user: User) -> None:
    if current_user.profile is None:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Onboarding precisa ser concluído antes de definir metas",
        )


@router.get("", response_model=CalorieGoalRead | None)
def get_current_goal(
    current_user: User = Depends(get_current_user), db: Session = Depends(get_db)
) -> CalorieGoal | None:
    return goal_service.get_current_goal(db, current_user.id)


@router.get("/suggestion", response_model=CalorieGoalAutoSuggestion)
def get_auto_suggestion(
    current_user: User = Depends(get_current_user), db: Session = Depends(get_db)
) -> dict:
    _require_profile(current_user)
    try:
        return goal_service.compute_suggestion(db, current_user.id, current_user.profile)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc)) from exc


@router.post("/auto", response_model=CalorieGoalRead)
def apply_auto_goal(
    as_first_objective: bool = False,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> CalorieGoal:
    """Aplica a meta automática. `as_first_objective=true` = "considerar como
    primeiro objetivo": o coach ESQUECE a fase anterior e vai direto pra meta
    nova (sem transição gradual) — e recomeça a análise a partir de agora (marco
    novo), pra não misturar cutting com bulking. Sem a flag, salto grande de kcal
    vira transição gradual."""
    _require_profile(current_user)
    try:
        suggestion = goal_service.compute_suggestion(db, current_user.id, current_user.profile)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc)) from exc
    goal = goal_service.apply_auto_goal(db, current_user.id, suggestion, immediate=as_first_objective)
    if as_first_objective:
        # Recomeça a leitura do coach a partir de agora — esquece a fase anterior.
        # Não apaga histórico (regra 4): só move o ponto de partida da análise.
        db.add(CoachingBaseline(user_id=current_user.id, effective_from=datetime.now(timezone.utc), reason="first_objective"))
        # E limpa os overlays de treino da fase anterior (subir carga, trocar
        # exercício, deload, técnica): eram do objetivo antigo. Sem isto, o treino
        # segue mostrando avisos que a análise nova nem enxerga — e às vezes
        # contraditórios (mandar subir a carga E trocar o mesmo exercício).
        coach_overlays.clear_training_overlays(db, current_user.id)
        db.commit()
    return goal


@router.post("/manual", response_model=CalorieGoalRead)
def apply_manual_goal(
    payload: CalorieGoalManualCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> CalorieGoal:
    return goal_service.apply_manual_goal(db, current_user.id, payload)

File: mobile/src/api/goals.ts
import { api } from "./client";

export type CalorieGoal = {
  id: number;
  mode: "manual" | "auto";
  kcal: number;
  protein_g: number;
  carbs_g: number;
  fat_g: number;
  fiber_g: number | null;
  sodium_mg: number | null;
  sugar_g: number | null;
  created_at: string;
};

export type CalorieGoalSuggestion = {
  kcal: number;
  protein_g: number;
  carbs_g: number;
  fat_g: number;
  current_goal: CalorieGoal | null;
  changed_significantly: boolean;
};

export async function getCurrentGoal(): Promise<CalorieGoal | null> {
  const { data } = await api.get<CalorieGoal | null>("/goals/calorie");
  return data;
}

export async function getAutoSuggestion(): Promise<CalorieGoalSuggestion> {
  const { data } = await api.get<CalorieGoalSuggestion>("/goals/calorie/suggestion");
  return data;
}

/** Aplica a meta automática. asFirstObjective = "considerar como primeiro
 * objetivo": vai direto pra meta nova (sem transição gradual) e recomeça a
 * análise do coach — pra trocar de fase (ex: cutting→bulking) sem arrasto. */
export async function applyAutoGoal(asFirstObjective = false): Promise<CalorieGoal> {
  const { data } = await api.post<CalorieGoal>("/goals/calorie/auto", null, {
    params: asFirstObjective ? { as_first_objective: true } : undefined,
  });
  return data;
}

export async function applyManualGoal(payload: {
  kcal: number;
  protein_g: number;
  carbs_g: number;
  fat_g: number;
  fiber_g?: number | null;
  sodium_mg?: number | null;
  sugar_g?: number | null;
}): Promise<CalorieGoal> {
  const { data } = await api.post<CalorieGoal>("/goals/calorie/manual", payload);
  return data;
}
